### PR TITLE
[Frontend][ONNX] Support ONNX Scan operator

### PR DESCRIPTION
**Pull request by shengxinhu**
support onnx scan operator

### CODE REVIEW

File: tests/python/frontend/onnx/test_forward.py [hunk 1]
```diff
@@ -5774,6 +5774,139 @@ def repeat(N, D):
     )
 
 
+@tvm.testing.parametrize_targets
```
Comment (AndrewZhaoLuo) [inline tests/python/frontend/onnx/test_forward.py:5777]: in unsupported_onnx_tests can you remove 

    "test_scan9_sum",
    "test_scan_sum",

And rerun `pytest ... test_onnx_nodes:test_scan_sum` `pytest ... test_onnx_nodes:test_scan9_sum`

Comment (shengxinhu) [reply tests/python/frontend/onnx/test_forward.py:5775]: I have remove them in unsupported_onnx_tests, but could not run test_onnx_nodes[test_scan_sum] and test_onnx_nodes[test_scan9_sum] in docker ci_gpu, it said not found.

Comment (AndrewZhaoLuo) [reply tests/python/frontend/onnx/test_forward.py:5775]: Interesting. Well the tests fail so maybe you can try running locally with the latest version of onnx and onnxruntime. 

We should pass the official tests to guarantee correctness with the onnx spec.

**Review by AndrewZhaoLuo — APPROVED** [review]
LGTM

File: python/tvm/relay/frontend/onnx.py [hunk 2]
```diff
@@ -3125,6 +3125,223 @@ def _impl_v1(cls, inputs, attr, params):
...
+
+        for i in range(num_scan_outputs):
+            name, _, _, _ = get_info(body.output[i + num_state_outputs])
+            output_node = infer_type(loop_outputs[i + num_state_outputs])
+            shape = list(get_const_tuple(output_node.checked_type.shape))
+            shape.insert(scan_output_axes[i], max_loop_count)
+            dtype = output_node.checked_type.dtype
+            scan_output_vars.append(_expr.var(name, shape=shape, dtype=dtype))
+            scan_output_init.append(_op.zeros(shape, dtype))
+
```
Comment (shengxinhu) [inline python/tvm/relay/frontend/onnx.py:3224]: It seems has a more elegant method(similar as operator Loop existing implementation):
scan_output_vars.append(_expr.var(name, shape=_ty.Any() * len(shape), dtype=dtype))
scan_output_init.append(_op.rehape(_expr.const(np.array[]).astype(dtype)), [0] + [1]*(len(shape)-1))
so the following _op.strided_slice could be removed, as the scan_output_init is empty.
But I did not make it work, could you take a look?

**Comment by mbrookhart** [comment]
@shengxinhu My apologies, I missed this one, I should have reviewed it three weeks ago, it seems. Any change you can rebase to resolve the conflicts?

**Review by mbrookhart — APPROVED** [review]

**Comment by masahi** [comment]
The change seems to have a lot of duplication with `Loop` converter. Can we clean things up?

**Comment by shengxinhu** [comment]
Hi @masahi , I didn't get your point of duplication with loop converter, could you point it out sepecifically?

**Comment by masahi** [comment]
I mean, just looking at the two implementations, it's obvious that you started by copy-pasting Loop converter (even comments are identical). Can we refactor common code into a helper function, or is it difficult to cleanly extract common bits?

This is not just for usual code-hygiene reason. I believe there are some issues in Loop converter in preserving static shape information. I've never seen a practical ONNX model with `Loop` op that can be successfully compiled by TVM - I've seen / heard that ONNX models converted from TF ssd mobilenet v1, MaskRCNN, and mlperf DLRM has `Loop` op, all of them failed to compile with TVM. So I don't want to see the Loop converter implementation in question duplicated for other ops.

See for example this issue regarding ssd mobilenet v1 https://github.com/apache/tvm/issues/8284#issuecomment-864467897 Note that the original tf ssd mobilenet v1 model compiles and runs fine when imported directly by TF frontend, so I can only assume that this is a ONNX converter issue.

**Comment by shengxinhu** [comment]
Hi @masahi , thanks for your explanations. I indeed copy some code in Loop converter implementation. I think only the get_var function could be extracted. do you think is it necessary? If Loop is questionable, Scan is likely to be the same. I also have not seen an ONNX model with scan operator, I realized it just for a task assigned to me.

**Comment by masahi** [comment]
ok, no problem.

**Pull request merged**